### PR TITLE
RUN-5108 adds ability for external windows to snap

### PR DESCRIPTION
**Pull request by whyn07m3**
This adds ability for an external window (only Notepad for now) to snap, and unsnap. This also shows potential targets while moving.

### CODE REVIEW

File: src/client/main.ts [hunk 1]
```diff
@@ -28,6 +28,11 @@ export interface WindowIdentity {
      * Window identifier
      */
     name: string;
+
+    /**
+     * External Window identifier
+     */
+    isExternalWindow?: boolean;
```
Comment (whyn07m3) [inline src/client/main.ts]: this property will differentiate identity of an OpenFin window from an External Window

File: src/provider/snapanddock/Projector.ts [hunk 2]
```diff
@@ -92,7 +92,11 @@ export class Projector {
             const snapOffset: Point = {x: 0, y: 0};
             const halfSize: Point = PointUtils.clone(activeState.halfSize);
             const validDirections: BorderProjection[] = borders.filter((border: BorderProjection) => {
-                return border.distance < Number.MAX_SAFE_INTEGER && border.getOverlap(activeState) >= MIN_OVERLAP && border.distance <= SNAP_DISTANCE;
```
Comment (whyn07m3) [inline src/provider/snapanddock/Projector.ts]: split this into separate parts to improve debugging

File: src/provider/snapanddock/Resolver.ts [hunk 3]
```diff
@@ -105,9 +105,12 @@ export class Resolver {
                     // Need to iterate over every window in both groups
                     activeGroup.entities.forEach(activeWindow => {
                         const activeState: EntityState = activeWindow.currentState;
+                        const isSnappable = this.isSnappable(activeWindow, activeState);
+                        const distance = RectUtils.distance(candidateGroup, activeState);
+                        const isDistanceWithin = distance.within(SNAP_DISTANCE);
 
                         // Only do the next loop if there's a chance that this window can intersect with the other group
-                        if (this.isSnappable(activeWindow, activeState) && RectUtils.distance(candidateGroup, activeState).within(SNAP_DISTANCE)) {
```
Comment (whyn07m3) [inline src/provider/snapanddock/Resolver.ts]: split this into separate parts to improve debugging

File: src/provider/snapanddock/Resolver.ts [hunk 4]
```diff
@@ -170,6 +173,10 @@ export class Resolver {
      * @param state State of the window object we are considering for snapping
      */
     private isSnappable(window: DesktopEntity, state: EntityState): boolean {
-        return !state.hidden && state.opacity > 0 && state.state === 'normal' && this._config.query(window.scope).features.snap;
```
Comment (whyn07m3) [inline src/provider/snapanddock/Resolver.ts]: split this into separate parts to improve debugging

File: src/provider/model/DesktopModel.ts [hunk 5]
```diff
@@ -18,6 +18,7 @@ import {DesktopTabGroup} from './DesktopTabGroup';
 import {DesktopWindow, EntityState, WindowIdentity} from './DesktopWindow';
 import {MouseTracker} from './MouseTracker';
 import {ZIndexer} from './ZIndexer';
+import { wrapWindow } from '../utils/main';
```
Comment (ryanpetersonOF) [inline src/provider/model/DesktopModel.ts]: Needs linting

File: src/provider/model/DesktopModel.ts [hunk 6]
```diff
@@ -88,6 +89,15 @@ export class DesktopModel {
                 });
             });
         });
+        (<any>fin.System).getAllExternalWindows().then((externalWindows: any) => {
```
Comment (ryanpetersonOF) [inline src/provider/model/DesktopModel.ts]: Do we not know what the type of `externalWindows` will be?  I think linting will flag this

Comment (whyn07m3) [reply src/provider/model/DesktopModel.ts]: All the shapes for external windows will be available in the new version of js-adapter, but that's in the future, right now i'm casting to `any` to pass linting.

File: src/provider/model/DesktopModel.ts (hunk 5 above)
Comment (whyn07m3) [reply src/provider/model/DesktopModel.ts]: done

**Pull request merged**